frontend/ui/widget/htmlboxwidget.lua

```diff
@@ -283,4 +461,78 @@ function HtmlBoxWidget:onTapText(arg, ges)
     end
 end
 
+function HtmlBoxWidget:setPageNumber(page_number)
+    if page_number == self.page_number then
+        return
+    end
+
+    self.page_number = page_number
+    self.page_boxes = nil
+    self:clearHighlight()
+end
+
+-- Returns true if the highlight has changed.
+function HtmlBoxWidget:clearHighlight()
+    self.hold_start_pos = nil
+    self.hold_end_pos = nil
+    return self:updateHighlight()
+end
+
+-- Returns true if the highlight has changed.
+function HtmlBoxWidget:updateHighlight()
+    if self.hold_start_pos and self.hold_end_pos then
+        -- getPageText is slow so we only call it when needed, and keep the result.
+        if self.page_boxes == nil then
+            local page = self.document:openPage(self.page_number)
+            self.page_boxes = page:getPageText()
+            page:close()
+        end
+
+        local text, rects = getSelectedText(self.page_boxes, self.hold_start_pos, self.hold_end_pos)
+        local changed = not areTextBoxesEqual(self.highlight_rects, self.highlight_text, rects, text)
+        if changed then
+            self.highlight_rects = rects
+            self.highlight_text = text
+        end
+        return changed
+    else
+        local changed = self.highlight_rects ~= nil
+        self.highlight_rects = nil
+        self.highlight_text = nil
+        return changed
+    end
+end
+
+function HtmlBoxWidget:redrawHighlight()
+    if self.highlight_text_selection then
+        self:freeBb()
+        UIManager:setDirty(self.dialog or "all", function()
+            return "ui", self.dimen
+        end)
+    end
+end
+
+function HtmlBoxWidget:scheduleClearHighlightAndRedraw()
+    if self.highlight_clear_and_redraw_function then
+        return
+    end
+
+    self.highlight_clear_and_redraw_function = function ()
+        self.highlight_clear_and_redraw_function = nil
```

Comment:
There is a little convention (not sure it's always followed) for these kind of scheduled function, we name them ending with `_action`, so possibly here `self.highlight_clear_and_redraw_action`.